### ISSUE TITLE
KLD Loss

### ISSUE THREAD

**Issue by zhl24** [yongkyung-oh/Stable-Neural-SDEs#8]
Hi, thanks for releasing the code and for the interesting discussion in your paper on the stability of Neural SDEs. It is an excellent code base with good tutorials, and I just have a question regarding the use of the KL divergence in training.  

From my own experience, when training with model evidence (e.g. reconstruction loss), unstable regimes are usually automatically avoided even without explicit stability constraints. Therefore, I was curious about the role of KL in your implementation.  

Looking into the code/tutorial, it seems that KL is computed from an additional neural network outputting `mu` and `logvar`, rather than from the process paths or model outputs directly. This confused me a bit, since in that case KL does not appear to interact with the data — the NN could simply learn to predict low KL values without affecting the dynamics. It is very possible that I have misunderstood something here. To check this, I experimented with removing the MSE part of the loss. In that setting, although the KL term decreases during training, the model predictions do not improve at all. (Figure below for reference.)  Once I re-introduce the MSE loss, training recovers as expected. 

<img width="595" height="993" alt="Image" src="https://github.com/user-attachments/assets/100abc08-024f-48d8-9352-60acb01a46ab" />

<img width="574" height="998" alt="Image" src="https://github.com/user-attachments/assets/e77364f8-0f90-41e0-ac22-0b88f0ee81f4" />

So my question is: **could you clarify the intended role of the KL loss in your training objective, and whether it is expected to meaningfully regularize the dynamics beyond the reconstruction loss?**  

Thanks again for the code release — I really appreciate it and I think a clarification here would help a lot of users.

**Comment by yongkyung-oh** [comment]
Hi, thanks for your question and for testing the code. You are right that the KL term comes from the auxiliary latent distribution, not the process paths. Its main role is to regularize the latent space, not to stabilize the dynamics. Without the reconstruction term, the KL loss alone cannot fit the data, so training collapses. The reconstruction loss fits the data, and the KL acts as a weak prior that smooths the latent space and prevents overfitting.

Furthermore, current tutorial is too simplified. It only generates data through integration and linear projection, which may be insufficient. You could add a decoder or generator for imputation or forecasting tasks.

Thanks again for raising this.

**Comment by zhl24** [comment]
> Hi, thanks for your question and for testing the code. You are right that the KL term comes from the auxiliary latent distribution, not the process paths. Its main role is to regularize the latent space, not to stabilize the dynamics. Without the reconstruction term, the KL loss alone cannot fit the data, so training collapses. The reconstruction loss fits the data, and the KL acts as a weak prior that smooths the latent space and prevents overfitting.
> 
> Furthermore, current tutorial is too simplified. It only generates data through integration and linear projection, which may be insufficient. You could add a decoder or generator for imputation or forecasting tasks.
> 
> Thanks again for raising this.

Hi, thanks for your reply. I see, so it seems like you are trying to use some rough approximation to the following evaluation of ELBO, and basically use a single generation likelihood as a rough Monte Carlo estimate to the first expectation, and then letting the first term dominates?

<img width="615" height="86" alt="Image" src="https://github.com/user-attachments/assets/cd3a56d6-eb15-452b-ba02-04e9b2f16f6a" />

Thanks again for the tutorial. It is very interesting and educational, and I have another question for the MLP implemented in the tutorials (in the screenshot). May I confirm if it satisfies both the Lipschitz and Growth conditions in the paper, since it is using scaled silu activation which is not mentioned in assumption 3.1 and 3.2 in the paper? It seems to me that a single such function does satisfy the Lipschitz and growth conditinos, but can these properties be kept after multiple layers? 
<img width="523" height="306" alt="Image" src="https://github.com/user-attachments/assets/0ea755a8-8cb2-46b7-a921-bf8988e091fa" />

**Comment by yongkyung-oh** [comment]
The tutorial computes a single-sample Monte Carlo estimate of the ELBO, where the reconstruction term dominates and the KL term serves only as a weak latent prior. It is meant as a simple demonstration rather than a full variational design.

The *scaled SiLU* (also called **LipSwish**) is used so that the activation has a global Lipschitz constant. This satisfies the paper’s Lipschitz and smoothness assumptions, which control the discriminator’s Lipschitz constant. 

References:
* [arXiv:2105.13493](https://arxiv.org/pdf/2105.13493)
* [torchsde example](https://github.com/google-research/torchsde/blob/master/examples/sde_gan.py)

For a multi-layer MLP, these properties are preserved because the overall Lipschitz constant is bounded by the product of layer spectral norms and the activation constant. With bounded weights and LipSwish (or any relevant activation), the network remains Lipschitz and satisfies the linear growth condition.

**Comment by zhl24** [comment]
Many thanks for helping out with the explanation. The paper explains things well and has pointed out several other interesting papers. It is really fascinating that the overall Lipschitz constant is bounded by the product of individual layers and <=Lipschitz-1 is needed for maintain stable exponential growth. Thanks again for your patience and help!
